Update celery router

iamsobanjaved opened the pull request:
Followed the new way of task routing explained in https://docs.celeryproject.org/en/v4.4.7/userguide/routing.html#routers

iamsobanjaved commented:
@jmbowman I have updated the router in this PR

iamsobanjaved commented:
Sandbox: https://iamsobanjaved.sandbox.edx.org/
Flower to monitor tasks: http://18.234.113.18:5555/

Now tasks on the sandbox are going to respective queues.

![image](https://user-images.githubusercontent.com/58461728/95611228-f8b79200-0a7a-11eb-947f-627e765ad0b4.png)

iamsobanjaved commented:
FYI @jmbowman @adzuci @cpennington

iamsobanjaved commented:
This is the documentation of how we tested celery queues in this PR
https://openedx.atlassian.net/wiki/spaces/TE/pages/1925939223/How+to+test+celery+queues+in+edx-platform

cpennington commented on cms/celery.py:
```diff
@@ -25,29 +25,29 @@
+    alternate_env_tasks = {
+        'completion_aggregator.tasks.update_aggregators': 'lms',
+        'openedx.core.djangoapps.content.block_structure.tasks.update_course_in_cache': 'lms',
+        'openedx.core.djangoapps.content.block_structure.tasks.update_course_in_cache_v2': 'lms',
```
Maybe not as part of this PR, but shouldn't this configuration be configuration, rather than code?

jmbowman replied:
True, having this and `explicit_queues` as Django settings would simplify the implementation here somewhat (and let more of the code for this routing function be shared between cms and lms.

iamsobanjaved replied:
Okay, I will move it there and refactor this function.

jmbowman commented on lms/celery.py:
```diff
@@ -25,14 +24,35 @@
+    # Defines the task -> alternate worker queue to be used when routing.
+    explicit_queues = {
+        'openedx.core.djangoapps.content.course_overviews.tasks.async_course_overview_update': {
+            'queue': settings.GRADES_DOWNLOAD_ROUTING_KEY},
```
Did these need to be added because `routing_key` on the `@task` decorator and `Task` objects is deprecated?  If so, should it also be removed from the decorators?

iamsobanjaved replied:
@jmbowman You have to mention the `queue` parameter in the `@task` decorator. We have verified this in this PR https://github.com/edx/edx-platform/pull/25314. We can either go by adding the `queue` parameter in the `@task` decorator or by removing `rotuing_key` from there and keep routing in `explicit_queues` dict. Which approach should I go with?

jmbowman replied:
I have a mild preference for keeping the queue assignments here (one step towards configuring this in settings rather than hard-coding), so we just nee to remove the `routing_key` attributes.

iamsobanjaved commented:
Moving `expicit_queues` to the Django settings file is causing some problems. Trying to identify the cause of the error by this step.

iamsobanjaved commented:
@jmbowman I moved `explicit_queues` and `alternate_env_tasks` to Django settings file `production.py` but I was getting some issues which were taking a lot of time to debug. So I reverted this PR to the previous state. We can merge this and do the cleanup work in follow-up PR. In that PR, the `routing_key` argument can be removed from the `@task` decorator and moved these `dict` to the Django settings file. By merging this PR we can reduce the load from the default queue in PROD.

jmbowman reviewed (approved):
👍 Sounds good

iamsobanjaved commented:
@adzuci When I build a sandbox and run any task I am seeing this exception. When I earlier made sandbox with these changes everything was fine. To resolve this exception, I have to run this in the platform on sandbox https://stackoverflow.com/a/42200761/12704761. Can you look into this, is this issue sandbox related?


![image](https://user-images.githubusercontent.com/58461728/96711100-74d4a280-13b6-11eb-9ba7-ca3ab7f4ab19.png)

iamsobanjaved commented:
@adzuci this is the sandbox in which I didn't follow any above-mentioned steps and it has the exception when tasks are triggered.
Sandbox: https://celeryrouter.sandbox.edx.org/
Flower: http://54.91.16.90:5555/dashboard

jmbowman commented:
You may want to file an SRE ticket for getting help with that.

iamsobanjaved commented:
> You may want to file an SRE ticket for getting help with that.

I have filed an SRE ticket for it already.
https://openedx.atlassian.net/browse/DOS-1644
